firmware/application/apps/ui_fileman.cpp

```diff
@@ -238,84 +327,96 @@ FileLoadView::FileLoadView(
-		destination_path = destination_path + buffer;
-		rename_file(get_selected_path(), destination_path);
-		load_directory_contents(current_path);
-		refresh_list();
-	});
-}
+void FileManagerView::on_rename() {
+	auto& entry = get_selected_entry();
 
-void FileManagerView::on_refactor(NavigationView& nav) {
-	text_prompt(nav, name_buffer, max_filename_length, [this](std::string& buffer) {
+	// Don't rename ".."
+	if (entry.path == parent_dir_path)
+		return;
 
-		std::string destination_path = current_path.string();
-		if (destination_path.back() != '/')//if the path is not ended with '/', add '/'
-			destination_path += '/';
-
-		auto selected_path = get_selected_path();
-		auto extension = selected_path.extension().string();
-
-		if(extension.empty()){// Is Dir
-			destination_path = destination_path + buffer;
-			extension_buffer = "";
-		}else{//is File
-			destination_path = destination_path + buffer + extension_buffer;
-		}
+	name_buffer = entry.path.filename().string();
 
-		rename_file(get_selected_path(), destination_path);  //rename the selected file
-
-		if (!extension.empty() && selected_path.string().back() != '/' && extension.substr(1) == "C16") { //substr(1) is for ignore the dot
-			// Rename its partner ( C16 <-> TXT ) file.
-			auto partner_file_path = selected_path.string().substr(0, selected_path.string().size() - 4) + ".TXT";
-			destination_path = destination_path.substr(0, destination_path.size() - 4) + ".TXT";
-			rename_file(partner_file_path, destination_path);
-		} else if (!extension.empty() && selected_path.string().back() != '/' && extension.substr(1) == "TXT") {
-			// If the file user choose is a TXT file.
-			auto partner_file_path = selected_path.string().substr(0, selected_path.string().size() - 4) + ".C16";
-			destination_path = destination_path.substr(0, destination_path.size() - 4) + ".C16";
-			rename_file(partner_file_path, destination_path);
-		}
+	uint32_t cursor_pos = (uint32_t)name_buffer.length();
+	if (auto pos = name_buffer.find_last_of("."); pos != name_buffer.npos)
+		cursor_pos = pos;
 
-		load_directory_contents(current_path);
-		refresh_list();
+	text_prompt(nav_, name_buffer, cursor_pos, max_filename_length,
+		[this, &entry](std::string& renamed) {
+			auto renamed_path = fs::path{ renamed };
+			rename_file(get_selected_full_path(), current_path / renamed_path);
 
-	});
+			auto has_partner = partner_file_prompt(nav_, entry.path, "Rename",
+				[this, renamed_path](const fs::path& partner, bool should_rename) mutable {
+					if (should_rename) {
+						auto new_name = renamed_path.replace_extension(partner.extension());
+						rename_file(current_path / partner, current_path / new_name);
+					}
+					reload_current();
+				}
+			);
 
+			if (!has_partner)
+				reload_current();
+		});
 }
 
 void FileManagerView::on_delete() {
-	delete_file(get_selected_path());
-	load_directory_contents(current_path);
-	refresh_list();
+	auto& entry = get_selected_entry();
+
+	// Don't delete ".."
```

Comment:
I don't love this. It's probably fine, but it would be better to prevent the menu from being selectable when this entry is selected instead so each handler doesn't need this test.